### ISSUE TITLE
CA1859 ignores possible change of overload resolution

### ISSUE THREAD

**Issue by sirduke** [dotnet/roslyn-analyzers#6687]
### Analyzer

**Diagnostic ID**: [CA1859](https://learn.microsoft.com/ru-ru/dotnet/fundamentals/code-analysis/quality-rules/ca1859): `Use concrete types when possible for improved performance`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 8 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: 8.0.100-preview.3.23178.7

### Describe the bug

The analyzer ignores possible change of overload resolution after changing type to a concrete one

### Steps To Reproduce

```
abstract class Base
{
    public virtual string M(object o) => "M(object)";
}

sealed class Derived : Base
{
    public void M(string s) {}
}

internal class C
{

    private readonly Base _a = new Derived(); //false CA1859. Applying the suggestion breaks compilation

    public void Trigger()
    {
        var s = _a.M("");
    }
}
```

### Expected behavior
No CA1859 reported

### Actual behavior

CA1859 is reported for field "_a"
